File: calsync.rb
#!/usr/bin/ruby
require 'net/https'
require 'uri'
require 'time'
require 'json'
require 'yaml'
require 'optparse'

class LbCal

    def initialize

        config = YAML.load_file(File.join(__dir__, 'config.yaml'))

        @client_id = config['client_id']
        @client_secret = config['client_secret']
        @device_code = ""

        @cal_id = config['cal_id']

        @access_token = config['access_token']
        @refresh_token = config['refresh_token']

        @work_cal_api = config['work_cal_api']
        @work_email = config['work_email']
        @cal_query = config['cal_query']

    end

    def saveTokens(a,r)

        config = YAML.load_file('config.yml')
        config['access_token'] = a
        config['refresh_token'] = r

        File.open('config.yml', 'w') do |f|
            YAML.dump(config, f)
        end

    end

    def getAuth
        uri = URI.parse("https://accounts.google.com/o/oauth2/device/code")

        data = {
            "client_id" => @client_id,
            "scope" => "https://www.googleapis.com/auth/calendar"

        }

        h = Net::HTTP.new(uri.host, uri.port)
        h.use_ssl = true
        req = Net::HTTP::Post.new(uri.path)
        req.set_form_data(data)

        res = h.request(req)


        d = JSON.parse( res.body)
        @device_code = d["device_code"]
        puts "Go here: #{d['verification_url']}"
        puts "Enter this code: #{d['user_code']}"
        puts "After authorizing, hit Enter"

        g = $stdin.gets

        self.getTokens
    end

    def getTokens
        uri = URI.parse("https://accounts.google.com/o/oauth2/token")

        data = {
                "client_id" => @client_id,
                "client_secret" => @client_secret,
                "code" => @device_code,
                "grant_type" => "http://oauth.net/grant_type/device/1.0"
        }


        h = Net::HTTP.new(uri.host, uri.port)
        h.use_ssl = true
        req = Net::HTTP::Post.new(uri.path)
        req.set_form_data(data)

        res = h.request(req)


        d = JSON.parse( res.body)
        # p d

        @access_token = d['access_token']
        @refresh_token = d['refresh_token']

        self.saveTokens(@access_token,@refresh_token)
    end

  	def refreshToken
        uri = URI.parse("https://accounts.google.com/o/oauth2/token")

        data = {
                "client_id" => @client_id,
                "client_secret" => @client_secret,
                "refresh_token" => @refresh_token,
                "grant_type" => "refresh_token"
        }


        h = Net::HTTP.new(uri.host, uri.port)
        h.use_ssl = true
        req = Net::HTTP::Post.new(uri.path)
        req.set_form_data(data)

        res = h.request(req)


        d = JSON.parse( res.body)
        
        # puts "refreshing token"
        # puts "ref: #{res.code}"

        @access_token = d['access_token']

        self.saveTokens(@access_token,@refresh_token)

    end

    def apiCall(url, method, data)
 	
        uri = URI.parse(url)
		
		h = Net::HTTP.new(uri.host, uri.port)
        h.use_ssl = true

      	case method
          	when "get"

          		data['access_token'] = @access_token
     			uri.query = URI.encode_www_form( data )
    	       	req = Net::HTTP::Get.new uri.request_uri

          	when "post"

          		params = {"access_token" => @access_token}
     			uri.query = URI.encode_www_form( params )
    	       	req = Net::HTTP::Post.new uri.request_uri
    	       	req.body = JSON.generate(data)
    	       	req.add_field("content-type","application/json")

          	when "delete"    

          		params = {"access_token" => @access_token}
     			uri.query = URI.encode_www_form( params )
    	       	req = Net::HTTP::Delete.new uri.request_uri

        end

        res = h.request(req)
        
        #p res.code
        #p res.body

        case res.code
            when "200"
            	d = JSON.parse( res.body)
            when "401"
    			self.refreshToken
            	d = self.apiCall(url,method,data)
            else
            	d = res
        end

        return d
    
    end

    def getOld()

    	data = {

			q: @cal_query,
			orderBy: "startTime",
			singleEvents: "true",
			timeMin: Time.new.xmlschema
		}
		url = "https://www.googleapis.com/calendar/v3/calendars/#{@cal_id}/events"
		res = self.apiCall(url,"get",data);

		return res['items']

    end

    def deleteOld(eid)
    	# puts "delete: #{eid}"
    	url = "https://www.googleapis.com/calendar/v3/calendars/#{@cal_id}/events/#{eid}"
    	res = self.apiCall(url,"delete",{});

    	# p res
    end 


    def createEvent(title,startDate,endDate)

    	data = {

    		"summary" => title,
    		"description" => "lbclass",
    		"start" => {"date"=>startDate},
    		"end" => {"date"=>endDate},
    		"reminders" => {"useDefault"=>"false"}
    	}

		url = "https://www.googleapis.com/calendar/v3/calendars/#{@cal_id}/events"
    	res = self.apiCall(url,"post",data);
    	
    end


    def createEvents(items)
        puts "> creating entries"
    	items.each do |i|

    		title = self.eventFormat(i)
    		startDate = Date.parse(i['start'])
    		endDate = Date.parse(i['end'])+1

    		self.createEvent(title,startDate,endDate)
           
    	end

    end



    def lbSchedule()

        uri = URI.parse(@work_cal_api)
        data = {
            "method"=>"userCalendar",
            "email"=>@work_email
        }

        puts "> get sched"

        h = Net::HTTP.new(uri.host, uri.port)
        req = Net::HTTP::Post.new(uri.path)
        req.set_form_data(data)
 
        res = h.request(req)

        d =  JSON.parse(res.body)

    end


    def eventFormat(sch)

        cnt = sch['cnt'].to_i

        crstitle = sch['gcal_title'].split(" - ")

        t = "[#{cnt}]#{crstitle[0]}"

        if sch['loc'] != "Austin, TX"
        t = "#{t}/#{sch['loc']}"
        end

        return t
    end

end


g = LbCal.new

if ARGV[0] == "setup"
    LbCal.new.getAuth()
else
	#===Remove Old
	old = g.getOld()
	if old
		old.each do |i|
			 g.deleteOld(i['id'])
		end
	end
	#===Put Latest
	sch = g.lbSchedule()
	if sch.length != 0
		g.createEvents(sch)
	end
end

File: calsync.js
const http = require('https');
const fs = require('fs');
const readline = require('readline');
const {google} = require('googleapis');
const googleAuth = require('google-auth-library');
const yaml = require('js-yaml');

const SCOPES = ['https://www.googleapis.com/auth/calendar'];
const TOKEN_DIR = (process.env.HOME || process.env.HOMEPATH ||
    process.env.USERPROFILE) + '/.gcal-credentials/';
const TOKEN_PATH = TOKEN_DIR + 'lb-calendar-sync.json';

const calendar = google.calendar('v3');

let calendarId;
let workCalApi;
let workEmail;
let calQuery;

function authorize(credentials, callback) {
  let clientSecret = credentials.installed.client_secret;
  let clientId = credentials.installed.client_id;
  let redirectUrl = credentials.installed.redirect_uris[0];
  let OAuth2 = google.auth.OAuth2;
  let oauth2Client = new OAuth2(clientId, clientSecret, redirectUrl);
  // Check if previously stored token.
  fs.readFile(TOKEN_PATH, (err, token) => {
    if (err) {
      getNewToken(oauth2Client, callback);
    } else {
      oauth2Client.credentials = JSON.parse(token);
      google.options({auth:oauth2Client});
      callback();
    }
  });
}

function getNewToken(oauth2Client, callback) {
  var authUrl = oauth2Client.generateAuthUrl({
    access_type: 'offline',
    scope: SCOPES
  });
  console.log('Authorize this app by visiting this url: ', authUrl);
  var rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  rl.question('Enter the code from that page here: ', (code) => {
    rl.close();
    oauth2Client.getToken(code, (err, token) => {
      if (err) {
        console.log('Error while trying to retrieve access token', err);
        return;
      }
      oauth2Client.credentials = token;
      storeToken(token);
      google.options({auth: oauth2Client});
      callback();
    });
  });
}

function storeToken(token) {
  try {
    fs.mkdirSync(TOKEN_DIR);
  } catch (err) {
    if (err.code != 'EEXIST') {
      throw err;
    }
  }
  fs.writeFile(TOKEN_PATH, JSON.stringify(token));
  console.log('Token stored to ' + TOKEN_PATH);
}

async function getOldGCalEvents () {
  let options = {
    calendarId: 'primary',
    // timeMin: (new Date()).toISOString(),
    singleEvents: true,
    orderBy: 'startTime',
    q: calQuery
  };
  console.log('> Getting Old Events');
  let response = await calendar.events.list(options);
  let events = response.data.items;
  
  if (events.length != 0) {
    console.log('> Deleting Old Events');
    let tasks = [];
    for (let i = 0; i < events.length; i++) {
      let event = events[i];
      await deleteGCalEvent(event.id);    
    }
  } 
  getWorkEvents();

}

async function deleteGCalEvent (eventId,callback) {
  let options = {
    calendarId: 'primary',
    eventId: eventId
  };
  return await calendar.events.delete(options);
}

async function getWorkEvents () {
  let email = workEmail;
  var options = {
    host: workCalHost,
    path: `${workCalApi}${email}`
  };
  http.get(options, (res) => {
    let str = '';
    res.on('data', (chunk) => {
      str += chunk;
    });
    res.on('end', () => {
      let lbData = JSON.parse(str);
      console.log('> Getting Work Events');
      createGCalEvents(lbData);
    });
  });
}

async function createGCalEvents (data) {
  console.log('> Creating Events')
  let events = [];

  for (let i = 0; i < data.length; i++) {

    let startDate;
    let endDate;
    let start;
    let end;

    let eventData = {
      summary: eventFormat(data[i]),
      description: calQuery,
      reminders: {
        'useDefault': false
      }
    }

    if (data[i].custom_dates) {

      data[i].custom_dates.forEach(date => {
        let event = {...eventData}
        event['start'] = {
          dateTime: new Date(date.start),
          timeZone: data[i].timezone
        }
        event['end'] = {
          dateTime: new  Date(date.end),
          timeZone: data[i].timezone
        }

        events.push(event)

      });



    } else {
      startDate = new Date(data[i].start);
      endDate = new Date(data[i].end);
      endDate.setDate(endDate.getDate()+1);
      start = startDate.toISOString().split('T')[0]
      end = endDate.toISOString().split('T')[0]

      let event = {...eventData}
      event['start'] = {date: start}
      event['end'] = {date: end}

      events.push(event)
    }

  }
  for (e of events) {
    // console.log(e)
    await calendar.events.insert({
      calendarId: 'primary',
      resource: e
    });
  }
}

function eventFormat(sch) {
  let count = sch['cnt'];
  let courseTitle = sch['gcal_title'].split(' - ');
  let newTitle = '';
  if (sch['custom_name'] != '') {
    newTitle = `[${count}] ${sch['custom_name']}`;
  } else {
    newTitle = `[${count}] ${courseTitle[0]}`;
  }
  newTitle = `${newTitle} : ${sch['loc']}`;
  return newTitle;
}

fs.readFile('client_secret.json', (err, content) => {
  if (err) {
    console.log('Error loading client secret file: ' + err);
    return;
  }

  let config = yaml.load(fs.readFileSync('config.yaml', 'utf8'));
  workCalHost = config.workCalHost;
  workCalApi = config.workCalApi;
  workEmail = config.workEmail;
  calQuery = config.calQuery;

  authorize(JSON.parse(content), getOldGCalEvents);
});
